src/nlp/arrow_writer.py

```diff
@@ -100,10 +102,37 @@ def _write_array_on_file(self, pa_array):
         self._num_bytes += pa_array.nbytes
         self.pa_writer.write_batch(pa_batch)
 
+    def _sort_fields(self, datatype):
+        sorted_fields = sorted(datatype, key=lambda x: x.name)
+        sorted_names = [x.name for x in sorted_fields]
+        return sorted_names, sorted_fields
+
     def write_on_file(self):
         """ Write stored examples
         """
-        if self.current_rows:
+        ext_cols = set(map(lambda x: x.name, (filter(
+            lambda x: isinstance(x.type, pa.PyExtensionType),
+            self._type
+        ))))
+        # pyarrow.lib.ArrowNotImplementedError: Sequence converter
+        # for type extension<arrow.py_extension_type> not implemented
+        # so, we must sequence them ourselves
+        if ext_cols and self.current_rows:
+            entries = []
+            for row in self.current_rows:
+                row_list = list(
+                    map(lambda x: pa.array([row[x]], self._type[x].type)
+                        if x not in ext_cols else row[x], self._sorted_names))
+                row = pa.RecordBatch.from_arrays(row_list, schema=self.schema)
+                row = pa.Table.from_batches([row])
+                entries.append(row)
+            entries = pa.concat_tables(entries)
+            # selecing the first row and column "image" and casting to numpy"
+            # we can implement mehtods for "to torch or others "
+            # print(entries.slice(0,1).column("image").chunk(0).to_numpy())
+            for b in entries.to_batches():
+                self.pa_writer.write_batch(b)
```

Comment:
This will make it write row by row, we have to make sure it doesn't result in bad compression and performance.

Did you try to make one single array with all the rows ? Then we could call `self._write_array_on_file(pa_array)` as what we do below.

Reply:
What's the error you're getting if you use the code below ?